### PR TITLE
TOTP Recovery Keys

### LINKED ISSUE #150: TOTP Recovery codes
Now that TOTP-based 2FA has been implemented, we need to implement recovery codes in order to provide users a way to avoid getting locked out of their accounts.

- [x] Generate a set of recovery codes upon TOTP being enabled. (GitHub gives 16)
- [x] Allow the user to use recovery codes to sign in in place of a TOTP code.
- [x] Consume the code upon use.
- [x] Clean up after TOTP is disabled.
- [x] Provide the user a way to generate new codes.

**Tests:**

- [ ] ~Users can log in with a recovery code~
- [ ] ~Recovery codes are consumed on use~
- [ ] ~User is displayed an error message upon using an invalid recovery code~
- [x] Users are assigned a set of recovery codes upon enabling TOTP
- [x] Recovery codes are removed when the user disables TOTP
- [x] Codes can be regenerated

### PR DESCRIPTION
**Pull request by raccube**
Closes #150

### CODE REVIEW

File: app/controllers/user_controller.rb [hunk 1]
```diff
@@ -175,13 +175,14 @@ def begin_export
...
-      current_user.otp_secret_key = User.otp_random_secret(26)
+      current_user.otp_secret_key = User.otp_random_secret(25)
       current_user.save
 
-      @provisioning_uri = current_user.provisioning_uri(nil, issuer: APP_CONFIG[:hostname])
       qr_code = RQRCode::QRCode.new(current_user.provisioning_uri("Retrospring:#{current_user.screen_name}", issuer: "Retrospring"))
 
       @qr_svg = qr_code.as_svg({offset: 4, module_size: 4, color: '000;fill:var(--primary)'}).html_safe
+    else
+      @recovery_code_count = TotpRecoveryCode.where(user_id: current_user.id).count
```
Comment (nilsding) [inline app/controllers/user_controller.rb]: would say `current_user.totp_recovery_codes.count` would be more natural

File: app/controllers/user_controller.rb [hunk 2]
```diff
@@ -190,19 +191,27 @@ def update_2fa
...
+      redirect_to edit_user_security_path
     end
-
-    redirect_to edit_user_security_path
   end
 
   def destroy_2fa
     current_user.otp_module = :disabled
     current_user.save!
+    TotpRecoveryCode.remove_all_for(current_user)
```
Comment (nilsding) [inline app/controllers/user_controller.rb]: likewise, why not `current_user.totp_recovery_codes.delete_all`?

**Review by nilsding — CHANGES_REQUESTED** [review]
some minor things

**Review by nilsding — APPROVED** [review]

**Review by pixeldesu — APPROVED** [review]
LGTM 👍

**Pull request merged**